src/library_dylink.js

```diff
@@ -530,8 +530,8 @@ var LibraryDylink = {
       // table and memory regions.  Later threads re-use the same table region
       // and can ignore the memory region (since memory is shared between
       // threads already).
-      var needsAllocation = !handle || !{{{ makeGetValue('handle', C_STRUCTS.dso.mem_allocated, 'i8') }}};
-      if (needsAllocation) {
+      var firstLoad = !handle || !{{{ makeGetValue('handle', C_STRUCTS.dso.mem_allocated, 'i8') }}};
```

Comment:
I guess it's the dso.mem_allocated variable that gets set by whichever thread loads the dso first? Should this load be atomic, or is there some other mechanism that prevents races here?

Reply:
I don't think so.  The handle allocation for dso handles already has a lock around it.  See the calling code here:

https://github.com/emscripten-core/emscripten/blob/ae2f8d8ab7bc13a2f683be5ff1a648ce700186b8/system/lib/libc/dynlink.c#L136-L149

Reply:
Ah, makes sense.
So... if we do what we discussed and end up creating mechanism where we expose another way for users to load libraries other than dlopen, that library will need to have its own synchronization mechanism too. Should we put a comment somewhere that this needs to be called under a lock?

Reply:
Add a comment yes.  Basically the `handle` parameter is optional, but if its passed then its assumed that access to that handle (a native data structure) is exclusive within this function.